### PR TITLE
Add local_zero_div to specialize phase

### LINKED ISSUE #4086: gradient of binary_crossentropy(sigmoid(), ...) doesn't correctly optimize if reshape before or after sigmoid
```python
import theano
import theano.tensor as T
from theano.tensor.nnet import binary_crossentropy, sigmoid
import numpy
a = theano.tensor.tensor4()

In [27]: T.grad(binary_crossentropy(sigmoid(a.reshape((-1, 1))), 1).sum(), a).eval({a: numpy.array([[[[30.]]]])})
Out[27]: array([[[[ nan]]]])

In [28]: T.grad(binary_crossentropy(sigmoid(a).reshape((-1, 1)), 1).sum(), a).eval({a: numpy.array([[[[30.]]]])})
Out[28]: array([[[[ nan]]]])

In [29]: T.grad(binary_crossentropy(sigmoid(a), 1).sum(), a).eval({a: numpy.array([[[[30.]]]])})
Out[29]: array([[[[-0.]]]])
```

### PR DESCRIPTION
**Pull request by lamblin**
Add a test that this is enough to solve the binary_crossentropy(reshape) case.
Follow-up to #4216, finally fixes #4086.

### CODE REVIEW

File: theano/tensor/nnet/tests/test_nnet.py [hunk 1]
```diff
@@ -1684,3 +1685,15 @@ def test_elu():
     for alpha in 1.5, 2, -1, -1.5, -2:
         y = elu(x, alpha).eval({x: X})
         utt.assert_allclose(y, numpy.where(X > 0, X, alpha * (numpy.exp(X) - 1)))
+
+
+def test_binary_crossentropy_reshape():
+    # Reported as https://github.com/Theano/Theano/issues/4086
+    a = tensor.tensor4('a')
+    c = binary_crossentropy(sigmoid(a.reshape((-1, 1))), 1).sum()
+    ga = theano.grad(c, a)
```
Comment (nouiz) [inline theano/tensor/nnet/tests/test_nnet.py]: I would add as a test the second case that had problems:

T.grad(binary_crossentropy(sigmoid(a).reshape((-1, 1)), 1).sum(), a).eval({a: numpy.array([[[[30.]]]])})

Comment (lamblin) [inline theano/tensor/nnet/tests/test_nnet.py]: OK.

Comment (lamblin) [inline theano/tensor/nnet/tests/test_nnet.py]: Done.

**Comment by nouiz** [comment]
ok to merge when travis pass.

**Pull request merged**